doc/plutus/architecture/constraints/index.rst

```diff
@@ -0,0 +1,294 @@
+          , slTxOutputs            :: Map TxOutRef ChainIndexTxOut
+          -- ^ Unspent outputs that the script may want to spend
+          , slOtherScripts         :: Map ValidatorHash Validator
+          -- ^ Validators of scripts other than "our script"
+          , slOtherData            :: Map DatumHash Datum
+          -- ^ Datums that we might need
+          , slPaymentPubKeyHashes  :: Map PaymentPubKeyHash PaymentPubKey
+          -- ^ Public keys that we might need
+          , slTypedValidator       :: Maybe (TypedValidator a)
+          -- ^ The script instance with the typed validator hash & actual compiled program
+          , slOwnPaymentPubKeyHash :: Maybe PaymentPubKeyHash
+          -- ^ The contract's payment public key hash, used for depositing tokens etc.
+          , slOwnStakePubKeyHash   :: Maybe StakePubKeyHash
+          -- ^ The contract's stake public key hash (optional)
+          } deriving stock (Show, Generic)
+            deriving anyclass (ToJSON, FromJSON)
+
+  mkTx
+      :: ( FromData (DatumType a)
+         , ToData (DatumType a)
+         , ToData (RedeemerType a)
+         )
+      => ScriptLookups a
+      -> TxConstraints (RedeemerType a) (DatumType a)
+      -> Either MkTxError UnbalancedTx
+
+Example
+-------
+
+Here's a complete example using constraints.
+We want to create a Plutus app that locks some Ada in a script output and splits them evenly between two recipients.
+
+Let's start by declaring the required imports.
+
+.. literalinclude:: BasicAppConstraints.hs
+   :start-after: BLOCK0
+   :end-before: BLOCK1
+
+We will then declare the ``SplitData`` datatype which describes the two recipients of the funds, and the total amount of the funds denoted in Ada.
+
+.. literalinclude:: BasicAppConstraints.hs
+   :start-after: BLOCK1
+   :end-before: BLOCK2
+
+Based on this datatype, we will start by creating a function which will generate constraints for:
+
+1- the validator script which will validate the spending of it's script output based on ``SplitData``
+
+2- the transaction we will create to spend these script outputs
+
+.. literalinclude:: BasicAppConstraints.hs
+   :start-after: BLOCK2
+   :end-before: BLOCK3
+
+We can then define the validator script.
+
+.. literalinclude:: BasicAppConstraints.hs
+   :start-after: BLOCK3
+   :end-before: BLOCK4
+
+With this validator script, we can create two types of transactions:
+
+1- transactions which lock Ada to the script address
+
+2- transactions which unlock Ada that has been locked to the script address
+
+The function below handles case 1.
+Note that we only show the transaction creation step, i.e. not the balance, sign and submit steps.
+
+.. literalinclude:: BasicAppConstraints.hs
+   :start-after: BLOCK4
+   :end-before: BLOCK5
+
+Then with case 2, to unlock the funds, we can use the same constraints as the validator script.
+
+.. literalinclude:: BasicAppConstraints.hs
+   :start-after: BLOCK5
+   :end-before: BLOCK6
+
+Rationale
```

Comment:
I think this should be the first section, before specification. It gives a meaning to everything else.